data/tutorials/language/0lg_00_basic_datatypes.md

```diff
@@ -0,0 +1,808 @@
+```ocaml
+# [||];;
+- : 'a array = [||]
+```
+Remember, `'a` ("alpha") is a type variable, a type parameter that will be replaced by another type.
+
+Like `string` and `bytes`, arrays support direct access, but the syntax is not the same.
+```ocaml
+# [| 'x'; 'y'; 'z' |].(2);;
+- : char = 'z'
+```
+
+Arrays are mutable, meaning they can't be extended or shortened, but each element may be updated.
+```ocaml
+# let letter = [| 'v'; 'x'; 'y'; 'z' |];;
+val letter : char array = [|'v'; 'x'; 'y'; 'z'|]
+
+# letter.(2) <- '3';;
+- : unit = ()
+
+# letter;;
+- : char array = [|'v'; 'x'; '3'; 'z'|]
+```
+
+Operations on arrays are provided by the [`Array`](/api/Array.html) module. There is a dedicated tutorial on Arrays.
+
+#### Lists
+
+As literals, lists are very much like arrays. Here are the same previous examples turned into lists.
+```ocaml
+# [ 0; 1; 2; 3; 4; 5 ];;
+- : int list = [0; 1; 2; 3; 4; 5]
+
+# [ 'x'; 'y'; 'z' ];;
+- : char list = ['x'; 'y'; 'z']
+
+# [ "foo"; "bar"; "baz" ];;
+- : string list = ["foo"; "bar"; "baz"]
+```
+
+Like arrays, lists are finite sequences of values of the same type. They are polymorphic, too. However, lists are extensible, immutable, and don't support direct access to all the values they contain. Lists play a central role in functional programming, so they have a [dedicated tutorial](/docs/lists).
+
+Operations on lists are provided by the [`List`](/api/List.html) module. The `List.append` function concatenates two lists. It can be used as an operator with the symbol `@`.
+
+There are symbols of special importance with respect to lists:
+- The empty list is written `[]`, has type `'a list'`, and is pronounced “nil.”
+- The list constructor operator, written `::`, and pronounced “cons,” is used to add a value at the head of a list.
+
+Together, they are the basic means to build a list and access the data it stores. For instance, here is how lists are built by successively applying the cons (`::`) operator:
+```ocaml
+# 3 :: [];;
+- : int list = [3]
+
+# 2 :: 3 :: [];;
+- : int list = [2; 3]
+
+# 1 :: 2 :: 3 :: [];;
+- : int list = [1; 2; 3]
+```
+
+Pattern matching provides the basic means to access data stored inside a list.
+```ocaml
+# match [1; 2; 3] with
+  | x :: u -> x
+  | [] -> raise Exit;;
+- : int = 1
+
+# match [1; 2; 3] with
+  | x :: y :: u -> y
+  | x :: u -> x
+  | [] -> raise Exit;;
+- : int = 2
+```
+
+In the above expressions, `[1; 2; 3]` is the value that is matched over. Each expression between the `|` and `->` symbols is a pattern. They are expressions of type list, only formed using `[]`, `::`, and variables names that represent various shapes a list may have. When the pattern is `[]`, it means “if the list is empty.” The pattern `x :: u` means “if the list contains data, let `x` be the first element of the list and `u` be the rest of the list.” Expressions at the right of the `->` symbol are the results returned in each corresponding case.
```

Comment:
@christinerose 
```suggestion
In the above expressions, `[1; 2; 3]` is the value that is matched over. Each expression between the `|` and `->` symbols is a pattern. They are expressions of type list, only formed using `[]`, `::`, and variables names that represent various shapes a list may have. Rhe pattern `[]` means “if the list is empty.” The pattern `x :: u` means “if the list contains data, let `x` be the first element of the list and `u` be the rest of the list.” Expressions at the right of the `->` symbol are the results returned in each corresponding case.
```

Should both sentences on both kind of patterns have the same structure? Both either with "when" or without?

Reply:
Nice! Yes!